golang/cosmos/x/swingset/keeper/snapshotter_test.go

```diff
@@ -0,0 +1,129 @@
+package keeper
+
+import (
+	"errors"
+	"testing"
+
+	"github.com/Agoric/agoric-sdk/golang/cosmos/vm"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/tendermint/tendermint/libs/log"
+)
+
+func newTestSnapshotter() SwingsetSnapshotter {
+	logger := log.NewNopLogger() // log.NewTMLogger(log.NewSyncWriter( /* os.Stdout*/ io.Discard)).With("module", "sdk/app")
+	return SwingsetSnapshotter{
+		isConfigured:      func() bool { return true },
+		takeSnapshot:      func(height int64) {},
+		newRestoreContext: func(height int64) sdk.Context { return sdk.Context{} },
+		logger:            logger,
+		blockingSend:      func(action vm.Jsonable) (string, error) { return "", nil },
+	}
+}
+
+func TestSnapshotInProgress(t *testing.T) {
+	swingsetSnapshotter := newTestSnapshotter()
+	swingsetSnapshotter.activeSnapshot = &activeSnapshot{}
+	err := swingsetSnapshotter.InitiateSnapshot(123)
+	if err == nil {
+		t.Error("wanted error for snapshot in progress")
+	}
+}
+
+func TestNotConfigured(t *testing.T) {
+	swingsetSnapshotter := newTestSnapshotter()
+	swingsetSnapshotter.isConfigured = func() bool { return false }
+	err := swingsetSnapshotter.InitiateSnapshot(123)
+	if err == nil {
+		t.Error("wanted error for unconfigured snapshot")
+	}
+}
+
+func TestSecondCommit(t *testing.T) {
+	swingsetSnapshotter := newTestSnapshotter()
+
+	// Use a channel to block the snapshot goroutine after it has started but before it exits.
+	ch := make(chan struct{}, 1)
+	swingsetSnapshotter.blockingSend = func(action vm.Jsonable) (string, error) {
+		ch <- struct{}{}
+		return "", nil
+	}
+
+	// First run through app.Commit()
+	err := swingsetSnapshotter.WaitUntilSnapshotStarted()
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = swingsetSnapshotter.InitiateSnapshot(123)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Second run through app.Commit() - should return right away
+	err = swingsetSnapshotter.WaitUntilSnapshotStarted()
```

Comment:
I made a note in #7349 that it's not practical to write tests that ensure that a goroutine is blocked - there's no "run until quiescent" command for the scheduler. (We asked Rob Pike for this in the early days, having had experience with nightly pseudorandom distributed system simulations tests for our Paxos system. Our request was denied.) The alternative is to put in a pause so that it's likely to be blocked if it hadn't finished - but that's flaky and makes the tests unnecessarily long.

So this test is meant to cover the 3rd bullet item from #7349, not the 2nd.

The test here allows the first controller call to complete immediately, which is why line 62 doesn't block. (The channel is created with a buffer of 1 on line 45).

Reply:
Oh yeah I definitely don't want flaky time based tests.

Can we at the very least test the internal structure used for the synchronization? Would the following work:

- call `InitiateSnapshot`
- verify that `activeSnapshotter` is not nil
- write in channel to unblock goroutine waiting on blockingSend
- call `WaitUntilSnapshotStarted`
- verify that `activeSnapshotter` is nil

Reply:
I think I've implemented what you want a slightly different way.